packages/flutter/lib/src/material/switch.dart

```diff
@@ -713,56 +741,101 @@ class _MaterialSwitchState extends State<_MaterialSwitch> with TickerProviderSta
+        ?? switchTheme.thumbColor?.resolve(activeStates);
+    final Color effectiveActiveThumbColor = activeThumbColor
+        ?? defaults.thumbColor!.resolve(activeStates)!;
+    final Color? inactiveThumbColor = widget.thumbColor?.resolve(inactiveStates)
+        ?? _widgetThumbColor.resolve(inactiveStates)
+        ?? switchTheme.thumbColor?.resolve(inactiveStates);
+    final Color effectiveInactiveThumbColor = inactiveThumbColor
+        ?? defaults.thumbColor!.resolve(inactiveStates)!;
     final Color effectiveActiveTrackColor = widget.trackColor?.resolve(activeStates)
-      ?? _widgetTrackColor.resolve(activeStates)
-      ?? switchTheme.trackColor?.resolve(activeStates)
-      ?? _defaultTrackColor.resolve(activeStates);
+        ?? _widgetTrackColor.resolve(activeStates)
+        ?? switchTheme.trackColor?.resolve(activeStates)
+        ?? _widgetThumbColor.resolve(activeStates)?.withAlpha(0x80)
+        ?? defaults.trackColor!.resolve(activeStates)!;
     final Color effectiveInactiveTrackColor = widget.trackColor?.resolve(inactiveStates)
-      ?? _widgetTrackColor.resolve(inactiveStates)
-      ?? switchTheme.trackColor?.resolve(inactiveStates)
-      ?? _defaultTrackColor.resolve(inactiveStates);
+        ?? _widgetTrackColor.resolve(inactiveStates)
+        ?? switchTheme.trackColor?.resolve(inactiveStates)
+        ?? defaults.trackColor!.resolve(inactiveStates)!;
+    final Color? effectiveInactiveTrackOutlineColor = switchConfig.trackOutlineColor?.resolve(inactiveStates);
+
+    final Widget? thumbWidget = widget.thumbImage?.resolve(states);
+    ImageProvider? effectiveActiveThumbImage;
+    ImageProvider? effectiveInactiveThumbImage;
+    if (thumbWidget is Image) {
+      effectiveActiveThumbImage = (widget.thumbImage?.resolve(activeStates) as Image?)?.image;
+      effectiveInactiveThumbImage = (widget.thumbImage?.resolve(inactiveStates) as Image?)?.image;
+    }
+    final Widget? thumbThemeWidget = switchTheme.thumbImage?.resolve(states);
+    effectiveActiveThumbImage ??= _widgetThumbImage.resolve(activeStates);
+    effectiveInactiveThumbImage ??= _widgetThumbImage.resolve(inactiveStates);
+    if (thumbThemeWidget is Image) {
+      effectiveActiveThumbImage ??= (switchTheme.thumbImage?.resolve(activeStates) as Image?)?.image;
+      effectiveInactiveThumbImage ??= (switchTheme.thumbImage?.resolve(inactiveStates) as Image?)?.image;
+    }
+
+    Icon? effectiveActiveIcon;
+    Icon? effectiveInactiveIcon;
+    if (theme.useMaterial3) {
+      if (thumbWidget is Icon) {
+        effectiveActiveIcon = widget.thumbImage?.resolve(activeStates) as Icon?;
+        effectiveInactiveIcon = widget.thumbImage?.resolve(inactiveStates) as Icon?;
+      }
+      if (thumbThemeWidget is Icon) {
+        effectiveActiveIcon ??= switchTheme.thumbImage?.resolve(activeStates) as Icon?;
+        effectiveInactiveIcon ??= switchTheme.thumbImage?.resolve(inactiveStates) as Icon?;
+      }
+    }
```

Comment:
Yeah, the `thumbImage` currently only supports `Image` and `Icon`. Investigating how we can paint a widget directly. Thanks for the suggestion!

Reply:
the thumb is currently a ShapeDecoration, so you can add a child to it somewhere.

Reply:
This is the challenge, because it uses a BoxDecoration and pre-renders/caches it. I don't know how to improve that.
`_cachedThumbPainter = _createDefaultThumbDecoration(thumbColor, thumbImage, thumbErrorListener).createBoxPainter(_handleDecorationChanged);
`

Reply:
Yes, this is tricky. But I agree that the restriction that only allows 2 types here is a little odd. To support the optional icon feature, I changed the `thumbImage` to the `thumbIcon`, and just used the type of `MaterialStateProperty<Icon?>?` instead of `MaterialStateProperty<Widget?>?`. Eventually, we will refactor this part of the code and allow a widget painting on the thumb of Switch.

Reply:
Perfect!